File: internal/routing/router_test.go
package routing

import (
	"context"
	"testing"

	"github.com/omniapi/omni-api/internal/model"
	"github.com/omniapi/omni-api/internal/upstream"
)

func fixtures() model.ExposedModel {
	config := model.Config{}
	for _, id := range []string{"one", "two", "three"} {
		config.Providers = append(config.Providers, model.Provider{
			ID:       id,
			Name:     id,
			BaseURL:  "https://example.test",
			Protocol: model.ProtocolOpenAIChat,
			Enabled:  true,
			Models:   []model.ProviderModel{{ID: id + "-model", UpstreamModel: "model", Alias: "chat"}},
		})
	}
	exposed, found := config.ExposedModel("chat")
	if !found {
		panic("fixture did not expose chat")
	}
	return exposed
}

func TestRoundRobinAdvancesStartingProvider(t *testing.T) {
	exposed := fixtures()
	router := NewRouter()
	calls := []string{}
	call := func(_ context.Context, target model.Target) (*upstream.Result, error) {
		calls = append(calls, target.Provider.ID)
		return &upstream.Result{Status: 200, Body: map[string]any{}}, nil
	}
	if _, err := router.Execute(context.Background(), exposed, call); err != nil {
		t.Fatalf("first call failed: %v", err)
	}
	if _, err := router.Execute(context.Background(), exposed, call); err != nil {
		t.Fatalf("second call failed: %v", err)
	}
	if len(calls) != 2 || calls[0] != "one" || calls[1] != "two" {
		t.Fatalf("expected [one two], got %v", calls)
	}
}

func TestSwitchesPastTwoFailures(t *testing.T) {
	exposed := fixtures()
	router := NewRouter()
	calls := []string{}
	result, err := router.Execute(context.Background(), exposed,
		func(_ context.Context, target model.Target) (*upstream.Result, error) {
			calls = append(calls, target.Provider.ID)
			if target.Provider.ID == "three" {
				return &upstream.Result{Status: 200, Body: map[string]any{}}, nil
			}
			return &upstream.Result{Status: 503, Body: map[string]any{}}, nil
		})
	if err != nil {
		t.Fatalf("expected success after failover, got %v", err)
	}
	if result.Status != 200 {
		t.Fatalf("expected status 200, got %d", result.Status)
	}
	if len(calls) != 3 || calls[2] != "three" {
		t.Fatalf("expected all three providers attempted in order, got %v", calls)
	}
}

func TestKeepsLastFailureWhenEveryProviderFails(t *testing.T) {
	exposed := fixtures()
	router := NewRouter()
	_, err := router.Execute(context.Background(), exposed,
		func(_ context.Context, target model.Target) (*upstream.Result, error) {
			if target.Provider.ID == "three" {
				return &upstream.Result{Status: 429, Body: map[string]any{}}, nil
			}
			return &upstream.Result{Status: 503, Body: map[string]any{}}, nil
		})
	failed, ok := err.(*AllProvidersFailedError)
	if !ok {
		t.Fatalf("expected AllProvidersFailedError, got %v", err)
	}
	if failed.Last.Status != 429 {
		t.Fatalf("expected the last failure status 429, got %d", failed.Last.Status)
	}
}

func TestDoesNotSwitchOnNonSwitchableStatus(t *testing.T) {
	exposed := fixtures()
	router := NewRouter()
	calls := []string{}
	_, err := router.Execute(context.Background(), exposed,
		func(_ context.Context, target model.Target) (*upstream.Result, error) {
			calls = append(calls, target.Provider.ID)
			return &upstream.Result{Status: 400, Body: map[string]any{}}, nil
		})
	if _, ok := err.(*AllProvidersFailedError); !ok {
		t.Fatalf("expected AllProvidersFailedError, got %v", err)
	}
	if len(calls) != 1 || calls[0] != "one" {
		t.Fatalf("expected a single attempt, got %v", calls)
	}
}

func TestRejectsWhenNoTargetExists(t *testing.T) {
	_, err := NewRouter().Execute(context.Background(), model.ExposedModel{Name: "chat"},
		func(context.Context, model.Target) (*upstream.Result, error) {
			t.Fatal("no provider should be called")
			return nil, nil
		})
	failed, ok := err.(*AllProvidersFailedError)
	if !ok {
		t.Fatalf("expected AllProvidersFailedError, got %v", err)
	}
	if failed.Last.Status != 503 {
		t.Fatalf("expected status 503, got %d", failed.Last.Status)
	}
}

func TestSkipsDisabledProviders(t *testing.T) {
	config := model.Config{Providers: []model.Provider{
		{ID: "off", Enabled: false, Models: []model.ProviderModel{{ID: "a", UpstreamModel: "model", Alias: "chat"}}},
		{ID: "on", Enabled: true, Models: []model.ProviderModel{{ID: "b", UpstreamModel: "model", Alias: "chat"}}},
	}}
	exposed, found := config.ExposedModel("chat")
	if !found {
		t.Fatal("expected chat to be exposed")
	}
	if len(exposed.Targets) != 1 || exposed.Targets[0].Provider.ID != "on" {
		t.Fatalf("expected only the enabled provider, got %+v", exposed.Targets)
	}
}

func TestAliasHidesUpstreamModelName(t *testing.T) {
	config := model.Config{Providers: []model.Provider{{
		ID:      "provider",
		Enabled: true,
		Models:  []model.ProviderModel{{ID: "a", UpstreamModel: "gpt-5.6-luna", Alias: "chat"}},
	}}}
	if _, found := config.ExposedModel("gpt-5.6-luna"); found {
		t.Fatal("aliased upstream model must not stay callable")
	}
	if _, found := config.ExposedModel("chat"); !found {
		t.Fatal("alias must be callable")
	}
}

File: tests/router.test.ts
import assert from 'node:assert/strict';
import test from 'node:test';
import { AllProvidersFailedError, RoundRobinRouter } from '../packages/core/src/router.js';
import type { LogicalModel, Provider } from '../packages/core/src/types.js';

const providers: Provider[] = ['one', 'two', 'three'].map((id) => ({
  id,
  name: id,
  baseUrl: 'https://example.test',
  protocol: 'openai-chat',
  enabled: true,
  models: [{ id: `${id}-model`, upstreamModel: 'model' }],
}));

const logicalModel: LogicalModel = {
  id: 'chat', name: 'chat', bindings: providers.map((provider) => ({
    providerId: provider.id, providerModelId: `${provider.id}-model`, enabled: true,
  })),
};

test('round robin advances the starting provider', async () => {
  const router = new RoundRobinRouter();
  const calls: string[] = [];
  const call = async (provider: Provider) => {
    calls.push(provider.id);
    return { status: 200, body: {} };
  };
  await router.execute(logicalModel, providers, call);
  await router.execute(logicalModel, providers, call);
  assert.deepEqual(calls, ['one', 'two']);
});

test('switches past two failures to a third provider', async () => {
  const router = new RoundRobinRouter();
  const calls: string[] = [];
  const result = await router.execute(logicalModel, providers, async (provider) => {
    calls.push(provider.id);
    return { status: provider.id === 'three' ? 200 : 503, body: {} };
  });
  assert.equal(result.status, 200);
  assert.deepEqual(calls, ['one', 'two', 'three']);
});

test('returns the last public upstream error when every provider fails', async () => {
  const router = new RoundRobinRouter();
  await assert.rejects(
    router.execute(logicalModel, providers, async (provider) => ({
      status: provider.id === 'three' ? 429 : 503, body: {},
    })),
    (error: AllProvidersFailedError) => error.lastFailure.status === 429,
  );
});

test('does not switch on a non-switchable 400 response', async () => {
  const router = new RoundRobinRouter();
  const calls: string[] = [];
  await assert.rejects(
    router.execute(logicalModel, providers, async (provider) => {
      calls.push(provider.id);
      return { status: 400, body: {} };
    }),
    AllProvidersFailedError,
  );
  assert.deepEqual(calls, ['one']);
});